swarm/src/behaviour.rs

```diff
@@ -250,6 +250,12 @@ pub enum ToSwarm<TOutEvent, TInEvent> {
     /// This allows a [`NetworkBehaviour`] to identify a connection that resulted out of its own dial request.
     Dial { opts: DialOpts },
 
+    /// Instructs the [`Swarm`](crate::Swarm) to listen on the provided address.
+    ListenOn { id: ListenerId, address: Multiaddr },
```

Comment:
This isn't quite consistent with `Dial` at the moment. Should we create a `ListenOpts` struct that by default creates the `ListenerId` for you and the user just has a ctor to pass in the `Multiaddr`?

Reply:
Either way would be fine to me, though If we did use `ListenOpts`, one might expect `Into<ListenOpts>` to be used for `Swarm::listen_on` as well, which would be fine to do, but would be redundant for it to return a `ListenerId` (and thus being a breaking change if its dropped from being returned).

Reply:
The entire PR is a breaking change anyway because we are adding variants to a public enum.

I am not sure if consistency with the `Swarm::dial` API is important enough here. I don't see us adding more options to `ListenOpts` any time soon, so it will always just be a `Multiaddr`.

For the `ToSwarm` case, having `ListenOpts` would be convenient because it allows users to construct the variant in one go, without having to also construct `ListenerId`.